pkg/kubelet/cm/internal_container_lifecycle.go

```diff
@@ -37,13 +38,22 @@ type internalContainerLifecycleImpl struct {
 	topologyManager topologymanager.Manager
 }
 
-func (i *internalContainerLifecycleImpl) PreStartContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
+func (i *internalContainerLifecycleImpl) PreCreateContainer(pod *v1.Pod, container *v1.Container, containerConfig *runtimeapi.ContainerConfig) error {
 	if i.cpuManager != nil {
-		err := i.cpuManager.AddContainer(pod, container, containerID)
-		if err != nil {
-			return err
+		allocatedCPUs := i.cpuManager.GetAllocatedResources(pod, container)
+		if allocatedCPUs != "" {
+			containerConfig.Linux.Resources.CpusetMems = allocatedCPUs
 		}
 	}
```

Comment:
Does something else protect this from getting called on non.linux systems? Is creation of the CPUManager somehow only limited to linux?

Reply:
In fact, I see this in `pkg/kubelet/cm/container_manager_windows.go`
```
func (cm *containerManagerImpl) InternalContainerLifecycle() InternalContainerLifecycle {
    return &internalContainerLifecycleImpl{cpumanager.NewFakeManager(), topologymanager.NewFakeManager()}
}
```
Which will obviously cause problems on windows since I'm guessing it doesn't support `containerConfig.Linux.Resources.CpusetMems`.

Reply:
We have two implementations of the container manager that compiled according to build flags, so if it windows specific build it will not include at all https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/internal_container_lifecycle.go

Reply:
I don't quite follow. The code I posted above clearly shows that the windows code instantiates an instance of `internalContainerLifecycleImpl`, passing it an instance of `cpumanager.NewFakeManager()`, meaning that `i.cpuManager != nil` will not be `nil` inside its call to `InternalContainerLifecycle()`.

Reply:
Oh, you are right, but in this case, FakeManager will return an empty string during the call to `GetAllocatedResources`, and the `if` check will not pass.

Reply:
Yeah. So it looks like it wouldn't cause issues in its current form, but it's also very subtle why it doesn't.
I would still either:
1. Add a `runtime.GOOS == "linux"` check here; or
2. Split the implementation of this one function into separate `_linux.go`, `_windows.go` files.

Reply:
I can see that we have the same thing for the `AddContainer` can we leave such refactoring to the separate PR or do you prefer to finish with it under the current one?

Reply:
`AddContainer()` is OK because it calls back into the `CPUManager `to do its thing. The real CPUManager will add the container to its internal containerMap and the fake CPUManager will do nothin (neither of which is linux specific). The only reason we need to branch for linux here is because we explicitly set a  linux specific field in `containerConfig.Linux.Resources.CpusetMems`.  If/when windows supports something similar, they can implement this differently using the same underlying CPUmanager.

Reply:
Makes sense, will create different files for Linux and windows implementation.